{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyPmJdW++r/PoOGkLAcIjc7W",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/haotianzh/scistreec/blob/main/scistreecna.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Check CUDA version and the number of CPU cores"
      ],
      "metadata": {
        "id": "p1BBmlxyjFry"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!nvcc --version\n",
        "!lscpu |grep 'CPU'"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qQR6Mc4PjK6W",
        "outputId": "4a0fd06a-e6b6-4add-8269-4d335cc48492"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "nvcc: NVIDIA (R) Cuda compiler driver\n",
            "Copyright (c) 2005-2024 NVIDIA Corporation\n",
            "Built on Thu_Jun__6_02:18:23_PDT_2024\n",
            "Cuda compilation tools, release 12.5, V12.5.82\n",
            "Build cuda_12.5.r12.5/compiler.34385749_0\n",
            "CPU op-mode(s):                          32-bit, 64-bit\n",
            "CPU(s):                                  2\n",
            "On-line CPU(s) list:                     0,1\n",
            "Model name:                              Intel(R) Xeon(R) CPU @ 2.20GHz\n",
            "CPU family:                              6\n",
            "NUMA node0 CPU(s):                       0,1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Install require packages, CUDA version is 12.5 here, so we use [cuda12x]"
      ],
      "metadata": {
        "id": "0pY2izErXFmY"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BZiQlYxaV2rt",
        "outputId": "24bd78e4-4c08-4a70-be7f-c0b75a87a6e7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'ScisTreeCNA'...\n",
            "remote: Enumerating objects: 192, done.\u001b[K\n",
            "remote: Counting objects: 100% (192/192), done.\u001b[K\n",
            "remote: Compressing objects: 100% (130/130), done.\u001b[K\n",
            "remote: Total 192 (delta 91), reused 152 (delta 54), pack-reused 0 (from 0)\u001b[K\n",
            "Receiving objects: 100% (192/192), 128.37 KiB | 5.83 MiB/s, done.\n",
            "Resolving deltas: 100% (91/91), done.\n",
            "Collecting ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git (from ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git)\n",
            "  Cloning https://github.com/haotianzh/ScisTreeCNA.git to /tmp/pip-install-apwmtpoj/scistreecna_fc2d705ada7348728f1a93d2ce5d9732\n",
            "  Running command git clone --filter=blob:none --quiet https://github.com/haotianzh/ScisTreeCNA.git /tmp/pip-install-apwmtpoj/scistreecna_fc2d705ada7348728f1a93d2ce5d9732\n",
            "  Resolved https://github.com/haotianzh/ScisTreeCNA.git to commit 77592af45aa161138445f685e50b851e5b2b5199\n",
            "  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (2.0.2)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.12/dist-packages (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (2.2.2)\n",
            "Collecting pptree (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git)\n",
            "  Downloading pptree-3.1.tar.gz (3.0 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting scistree2 (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git)\n",
            "  Downloading scistree2-0.4.0-py3-none-any.whl.metadata (9.4 kB)\n",
            "Collecting phytreeviz (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git)\n",
            "  Downloading phytreeviz-0.2.0-py3-none-any.whl.metadata (7.6 kB)\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.12/dist-packages (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (1.6.1)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.12/dist-packages (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (1.16.3)\n",
            "Collecting biopython (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git)\n",
            "  Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: rich in /usr/local/lib/python3.12/dist-packages (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (13.9.4)\n",
            "Requirement already satisfied: cupy-cuda12x in /usr/local/lib/python3.12/dist-packages (from ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (13.6.0)\n",
            "Requirement already satisfied: fastrlock>=0.5 in /usr/local/lib/python3.12/dist-packages (from cupy-cuda12x->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (0.8.3)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.12/dist-packages (from pandas->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.12/dist-packages (from pandas->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.12/dist-packages (from pandas->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (2025.2)\n",
            "Requirement already satisfied: matplotlib>=3.5.3 in /usr/local/lib/python3.12/dist-packages (from phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (3.10.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.12/dist-packages (from rich->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (4.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.12/dist-packages (from rich->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (2.19.2)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.12/dist-packages (from scikit-learn->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (1.5.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.12/dist-packages (from scikit-learn->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (3.6.0)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.12/dist-packages (from markdown-it-py>=2.2.0->rich->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (0.1.2)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.5.3->phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (1.3.3)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.5.3->phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.5.3->phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (4.60.1)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.5.3->phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (1.4.9)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.5.3->phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (25.0)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.5.3->phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (11.3.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.5.3->phytreeviz->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (3.2.5)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.12/dist-packages (from python-dateutil>=2.8.2->pandas->ScisTreeCNA@ git+https://github.com/haotianzh/ScisTreeCNA.git->ScisTreeCNA[cuda12x]@ git+https://github.com/haotianzh/ScisTreeCNA.git) (1.17.0)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m51.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading phytreeviz-0.2.0-py3-none-any.whl (20 kB)\n",
            "Downloading scistree2-0.4.0-py3-none-any.whl (732 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m732.1/732.1 kB\u001b[0m \u001b[31m55.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hBuilding wheels for collected packages: ScisTreeCNA, pptree\n",
            "  Building wheel for ScisTreeCNA (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for ScisTreeCNA: filename=scistreecna-0.1.0-py3-none-any.whl size=58211 sha256=4889aef3a62cf33f3646da6f257a178ecc20895f231de2d9860aab3b10c6ad1e\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-tb97q1z3/wheels/16/64/0a/f41a2141628d713817dde7f12b257c90c81daf413450c0cffd\n",
            "  Building wheel for pptree (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pptree: filename=pptree-3.1-py3-none-any.whl size=4609 sha256=ec9041ec378ac866468e0d3cb310cd3689d2713ab39e59b6f8fe39ab8c4a59dc\n",
            "  Stored in directory: /root/.cache/pip/wheels/b0/2d/de/37058114a8f07cfec75747cb46b864bc5c71b0e9e0e4cd0acd\n",
            "Successfully built ScisTreeCNA pptree\n",
            "Installing collected packages: pptree, biopython, phytreeviz, scistree2, ScisTreeCNA\n",
            "Successfully installed ScisTreeCNA-0.1.0 biopython-1.86 phytreeviz-0.2.0 pptree-3.1 scistree2-0.4.0\n"
          ]
        }
      ],
      "source": [
        "!git clone https://github.com/haotianzh/ScisTreeCNA.git\n",
        "!pip install \"ScisTreeCNA[cuda12x] @ git+https://github.com/haotianzh/ScisTreeCNA.git\""
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# New Section"
      ],
      "metadata": {
        "id": "_t9gOCb_mJ3X"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import pickle\n",
        "import scistreecna as scna"
      ],
      "metadata": {
        "id": "8JZWNEwmY0Wm"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Load example data"
      ],
      "metadata": {
        "id": "KBz7m0u3ZPTj"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "We begin by taking a quick look at the input data. Each row corresponds to a SNP site, and each column corresponds to a cell. In this example, the dataset contains **100 sites** and **60 cells**.\n",
        "\n",
        "For each *(cell, site)* pair, the entry is a string in the format **`ref|alt|cn`**, where:\n",
        "\n",
        "- **ref**: read count of the reference (wild-type) allele  \n",
        "- **alt**: read count of the mutant allele  \n",
        "- **cn**: observed copy number (either absolute copy number — recommended — or relative copy state)\n",
        "\n",
        "Missing values should be encoded as:\n",
        "\n",
        "- `.|.|cn` — if read counts are missing but copy number is available  \n",
        "- `.|.|.` — if both read counts and copy number are missing\n",
        "- `ref|alt|.` - if only copy number is missing"
      ],
      "metadata": {
        "id": "MrvYWg3zZUVT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "data = pd.read_csv('./ScisTreeCNA/examples/test_data_reads.csv', index_col=0)\n",
        "data"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 444
        },
        "id": "AVIC1ro0ZNy-",
        "outputId": "7108e895-2d53-4b45-a2c9-206c2504edfd"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "         c0      c1      c2      c3      c4      c5      c6      c7      c8  \\\n",
              "s0   11|0|2  20|1|1  25|0|2   9|0|2  36|0|2  17|1|2  13|0|1  27|1|2   7|0|2   \n",
              "s1    4|0|2  23|0|1  18|0|2  14|2|2  32|0|2   5|1|2   0|0|1   5|0|2  27|0|2   \n",
              "s2   13|0|2  26|0|2  20|0|2   6|0|2  17|0|2  12|0|2  34|0|2  21|0|2  24|0|2   \n",
              "s3   15|1|3  12|0|3  15|0|2   3|0|2  20|0|3  24|0|2  22|0|2  15|0|1  20|0|2   \n",
              "s4   22|0|2  15|0|2  15|5|2  19|0|2   9|0|2  17|7|2  24|0|2  19|0|2  10|0|2   \n",
              "..      ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
              "s95  20|0|2   5|0|2  11|0|2  10|0|2  16|0|2  10|0|2  12|0|2  19|0|2  13|0|2   \n",
              "s96  20|0|1  23|0|2  13|0|2  32|0|2  33|0|3   0|0|2   9|0|2   8|0|2  24|0|2   \n",
              "s97  19|0|2  12|7|2  28|0|2  28|0|2   8|0|2  12|0|2  33|0|2  25|0|1   7|0|2   \n",
              "s98  30|0|2  17|0|2  19|0|2  13|0|2  26|1|2   1|0|2   4|0|2   9|0|2  35|0|2   \n",
              "s99  20|0|2  24|1|2  29|0|2  15|0|2  12|0|2  11|0|2  35|0|2   8|0|2  32|0|2   \n",
              "\n",
              "         c9  ...     c50      c51     c52     c53     c54     c55     c56  \\\n",
              "s0   21|0|2  ...  23|0|2    5|0|2  22|1|2  13|0|2  34|1|2  15|0|2   0|0|2   \n",
              "s1    8|0|2  ...   9|0|2   11|1|2  14|0|2  13|0|2  17|0|2  15|0|2  14|0|2   \n",
              "s2   24|0|2  ...  11|0|1   12|2|2  13|0|2  22|0|2  21|1|2  20|0|2   2|1|2   \n",
              "s3   16|0|2  ...  40|0|3   25|0|2  19|0|2  38|0|3   4|0|2  12|0|2  43|0|3   \n",
              "s4    5|8|2  ...  22|0|2    6|4|2  32|0|2   4|0|2  26|1|2   6|5|2  10|7|2   \n",
              "..      ...  ...     ...      ...     ...     ...     ...     ...     ...   \n",
              "s95  25|0|2  ...  23|1|1  10|16|2   3|0|2  13|0|1  22|0|2  33|0|2  28|2|2   \n",
              "s96   9|0|2  ...  21|0|2    5|0|2   6|0|2  16|0|2   5|0|2  30|0|2   5|0|2   \n",
              "s97  11|0|2  ...  18|0|2   10|0|2   3|0|2  29|0|2   5|0|2  21|0|2  11|0|2   \n",
              "s98   0|0|2  ...  20|0|2   16|1|2  22|0|2   8|0|2  26|0|2   9|0|2  14|1|2   \n",
              "s99  18|0|2  ...  26|1|2   11|0|2  14|0|2  18|0|2  26|0|2  13|0|2  22|0|2   \n",
              "\n",
              "        c57     c58     c59  \n",
              "s0   23|0|2  12|0|2  18|0|2  \n",
              "s1   17|0|2  15|0|2   1|0|2  \n",
              "s2   21|1|2  13|0|2  16|0|2  \n",
              "s3   13|0|3  15|0|2  23|0|2  \n",
              "s4   23|0|2  25|0|2  10|0|2  \n",
              "..      ...     ...     ...  \n",
              "s95  30|0|2  14|0|1   0|0|2  \n",
              "s96   0|0|2   2|0|2  12|0|2  \n",
              "s97  35|0|1  25|1|2   4|0|2  \n",
              "s98  29|0|2  39|0|3  29|0|2  \n",
              "s99   0|0|2  0|13|2  18|1|2  \n",
              "\n",
              "[100 rows x 60 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-0eae9690-0b88-44e2-b66b-bec5e8fb9fc8\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>c0</th>\n",
              "      <th>c1</th>\n",
              "      <th>c2</th>\n",
              "      <th>c3</th>\n",
              "      <th>c4</th>\n",
              "      <th>c5</th>\n",
              "      <th>c6</th>\n",
              "      <th>c7</th>\n",
              "      <th>c8</th>\n",
              "      <th>c9</th>\n",
              "      <th>...</th>\n",
              "      <th>c50</th>\n",
              "      <th>c51</th>\n",
              "      <th>c52</th>\n",
              "      <th>c53</th>\n",
              "      <th>c54</th>\n",
              "      <th>c55</th>\n",
              "      <th>c56</th>\n",
              "      <th>c57</th>\n",
              "      <th>c58</th>\n",
              "      <th>c59</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>s0</th>\n",
              "      <td>11|0|2</td>\n",
              "      <td>20|1|1</td>\n",
              "      <td>25|0|2</td>\n",
              "      <td>9|0|2</td>\n",
              "      <td>36|0|2</td>\n",
              "      <td>17|1|2</td>\n",
              "      <td>13|0|1</td>\n",
              "      <td>27|1|2</td>\n",
              "      <td>7|0|2</td>\n",
              "      <td>21|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>23|0|2</td>\n",
              "      <td>5|0|2</td>\n",
              "      <td>22|1|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>34|1|2</td>\n",
              "      <td>15|0|2</td>\n",
              "      <td>0|0|2</td>\n",
              "      <td>23|0|2</td>\n",
              "      <td>12|0|2</td>\n",
              "      <td>18|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s1</th>\n",
              "      <td>4|0|2</td>\n",
              "      <td>23|0|1</td>\n",
              "      <td>18|0|2</td>\n",
              "      <td>14|2|2</td>\n",
              "      <td>32|0|2</td>\n",
              "      <td>5|1|2</td>\n",
              "      <td>0|0|1</td>\n",
              "      <td>5|0|2</td>\n",
              "      <td>27|0|2</td>\n",
              "      <td>8|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>9|0|2</td>\n",
              "      <td>11|1|2</td>\n",
              "      <td>14|0|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>17|0|2</td>\n",
              "      <td>15|0|2</td>\n",
              "      <td>14|0|2</td>\n",
              "      <td>17|0|2</td>\n",
              "      <td>15|0|2</td>\n",
              "      <td>1|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s2</th>\n",
              "      <td>13|0|2</td>\n",
              "      <td>26|0|2</td>\n",
              "      <td>20|0|2</td>\n",
              "      <td>6|0|2</td>\n",
              "      <td>17|0|2</td>\n",
              "      <td>12|0|2</td>\n",
              "      <td>34|0|2</td>\n",
              "      <td>21|0|2</td>\n",
              "      <td>24|0|2</td>\n",
              "      <td>24|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>11|0|1</td>\n",
              "      <td>12|2|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>22|0|2</td>\n",
              "      <td>21|1|2</td>\n",
              "      <td>20|0|2</td>\n",
              "      <td>2|1|2</td>\n",
              "      <td>21|1|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>16|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s3</th>\n",
              "      <td>15|1|3</td>\n",
              "      <td>12|0|3</td>\n",
              "      <td>15|0|2</td>\n",
              "      <td>3|0|2</td>\n",
              "      <td>20|0|3</td>\n",
              "      <td>24|0|2</td>\n",
              "      <td>22|0|2</td>\n",
              "      <td>15|0|1</td>\n",
              "      <td>20|0|2</td>\n",
              "      <td>16|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>40|0|3</td>\n",
              "      <td>25|0|2</td>\n",
              "      <td>19|0|2</td>\n",
              "      <td>38|0|3</td>\n",
              "      <td>4|0|2</td>\n",
              "      <td>12|0|2</td>\n",
              "      <td>43|0|3</td>\n",
              "      <td>13|0|3</td>\n",
              "      <td>15|0|2</td>\n",
              "      <td>23|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s4</th>\n",
              "      <td>22|0|2</td>\n",
              "      <td>15|0|2</td>\n",
              "      <td>15|5|2</td>\n",
              "      <td>19|0|2</td>\n",
              "      <td>9|0|2</td>\n",
              "      <td>17|7|2</td>\n",
              "      <td>24|0|2</td>\n",
              "      <td>19|0|2</td>\n",
              "      <td>10|0|2</td>\n",
              "      <td>5|8|2</td>\n",
              "      <td>...</td>\n",
              "      <td>22|0|2</td>\n",
              "      <td>6|4|2</td>\n",
              "      <td>32|0|2</td>\n",
              "      <td>4|0|2</td>\n",
              "      <td>26|1|2</td>\n",
              "      <td>6|5|2</td>\n",
              "      <td>10|7|2</td>\n",
              "      <td>23|0|2</td>\n",
              "      <td>25|0|2</td>\n",
              "      <td>10|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s95</th>\n",
              "      <td>20|0|2</td>\n",
              "      <td>5|0|2</td>\n",
              "      <td>11|0|2</td>\n",
              "      <td>10|0|2</td>\n",
              "      <td>16|0|2</td>\n",
              "      <td>10|0|2</td>\n",
              "      <td>12|0|2</td>\n",
              "      <td>19|0|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>25|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>23|1|1</td>\n",
              "      <td>10|16|2</td>\n",
              "      <td>3|0|2</td>\n",
              "      <td>13|0|1</td>\n",
              "      <td>22|0|2</td>\n",
              "      <td>33|0|2</td>\n",
              "      <td>28|2|2</td>\n",
              "      <td>30|0|2</td>\n",
              "      <td>14|0|1</td>\n",
              "      <td>0|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s96</th>\n",
              "      <td>20|0|1</td>\n",
              "      <td>23|0|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>32|0|2</td>\n",
              "      <td>33|0|3</td>\n",
              "      <td>0|0|2</td>\n",
              "      <td>9|0|2</td>\n",
              "      <td>8|0|2</td>\n",
              "      <td>24|0|2</td>\n",
              "      <td>9|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>21|0|2</td>\n",
              "      <td>5|0|2</td>\n",
              "      <td>6|0|2</td>\n",
              "      <td>16|0|2</td>\n",
              "      <td>5|0|2</td>\n",
              "      <td>30|0|2</td>\n",
              "      <td>5|0|2</td>\n",
              "      <td>0|0|2</td>\n",
              "      <td>2|0|2</td>\n",
              "      <td>12|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s97</th>\n",
              "      <td>19|0|2</td>\n",
              "      <td>12|7|2</td>\n",
              "      <td>28|0|2</td>\n",
              "      <td>28|0|2</td>\n",
              "      <td>8|0|2</td>\n",
              "      <td>12|0|2</td>\n",
              "      <td>33|0|2</td>\n",
              "      <td>25|0|1</td>\n",
              "      <td>7|0|2</td>\n",
              "      <td>11|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>18|0|2</td>\n",
              "      <td>10|0|2</td>\n",
              "      <td>3|0|2</td>\n",
              "      <td>29|0|2</td>\n",
              "      <td>5|0|2</td>\n",
              "      <td>21|0|2</td>\n",
              "      <td>11|0|2</td>\n",
              "      <td>35|0|1</td>\n",
              "      <td>25|1|2</td>\n",
              "      <td>4|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s98</th>\n",
              "      <td>30|0|2</td>\n",
              "      <td>17|0|2</td>\n",
              "      <td>19|0|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>26|1|2</td>\n",
              "      <td>1|0|2</td>\n",
              "      <td>4|0|2</td>\n",
              "      <td>9|0|2</td>\n",
              "      <td>35|0|2</td>\n",
              "      <td>0|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>20|0|2</td>\n",
              "      <td>16|1|2</td>\n",
              "      <td>22|0|2</td>\n",
              "      <td>8|0|2</td>\n",
              "      <td>26|0|2</td>\n",
              "      <td>9|0|2</td>\n",
              "      <td>14|1|2</td>\n",
              "      <td>29|0|2</td>\n",
              "      <td>39|0|3</td>\n",
              "      <td>29|0|2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>s99</th>\n",
              "      <td>20|0|2</td>\n",
              "      <td>24|1|2</td>\n",
              "      <td>29|0|2</td>\n",
              "      <td>15|0|2</td>\n",
              "      <td>12|0|2</td>\n",
              "      <td>11|0|2</td>\n",
              "      <td>35|0|2</td>\n",
              "      <td>8|0|2</td>\n",
              "      <td>32|0|2</td>\n",
              "      <td>18|0|2</td>\n",
              "      <td>...</td>\n",
              "      <td>26|1|2</td>\n",
              "      <td>11|0|2</td>\n",
              "      <td>14|0|2</td>\n",
              "      <td>18|0|2</td>\n",
              "      <td>26|0|2</td>\n",
              "      <td>13|0|2</td>\n",
              "      <td>22|0|2</td>\n",
              "      <td>0|0|2</td>\n",
              "      <td>0|13|2</td>\n",
              "      <td>18|1|2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>100 rows × 60 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0eae9690-0b88-44e2-b66b-bec5e8fb9fc8')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-0eae9690-0b88-44e2-b66b-bec5e8fb9fc8 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-0eae9690-0b88-44e2-b66b-bec5e8fb9fc8');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-5d900ee2-585f-452c-bd97-e10628b5e25f\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-5d900ee2-585f-452c-bd97-e10628b5e25f')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-5d900ee2-585f-452c-bd97-e10628b5e25f button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "  <div id=\"id_85b13144-c24c-403b-bdee-d6584f9c3499\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('data')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_85b13144-c24c-403b-bdee-d6584f9c3499 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('data');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "data"
            }
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, we load the data and parse it into a NumPy `ndarray`. We also extract the `cell_names` and `site_names` from the input table.  \n",
        ">Note: Users may also prepare this data array directly from their own data, without explicitly creating a `.csv` file as shown above. ScisTreeCNA works directly with a `numpy.ndarray`.\n"
      ],
      "metadata": {
        "id": "L35cpfvrZiz2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# load .csv file\n",
        "reads, cell_names, site_names = scna.util.read_csv('./ScisTreeCNA/examples/test_data_reads.csv')"
      ],
      "metadata": {
        "id": "Ah0rvg5hZmRe"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now we are ready to run **scistreecna**. Before doing so, we need to configure several parameters, including allelic dropout (ADO), sequencing error rate, and copy-number noise.  \n",
        "We must also specify the expected **minimum** and **maximum** copy numbers that best match the dataset. The minimum should be **greater than** 0, although the observed copy number in the input data may be 0.\n",
        "\n",
        "In addition, `tree_batch_size` and `node_batch_size` control how the GPU parallelizes the computation. Setting these values too high may lead to out-of-memory (OOM) errors.\n"
      ],
      "metadata": {
        "id": "ADA7DvCIZs9y"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# run inferenceb\n",
        "scistreecna_tree, scistreecna_geno = scna.infer(reads,\n",
        "                                                cell_names=cell_names,  # cell names\n",
        "                                                ado=0.1,  # allelic dropout rate\n",
        "                                                seq_error=0.01,   # sequencing error\n",
        "                                                cn_noise=0.05,    # copy number noise\n",
        "                                                cn_min=1, # minimum copy number (>=1)\n",
        "                                                cn_max=5, # maximum copy number\n",
        "                                                tree_batch_size=128,  # number of trees evaluated together\n",
        "                                                node_batch_size=256,  # number of nodes evaluated together\n",
        "                                                verbose=True)  # print logs"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KMDnpb13ZyH-",
        "outputId": "05a1b9aa-bced-4871-8409-097a013b94d0"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "───────────────────────────────────────── ScisTreeCNA ──────────────────────────────────────────\n",
            "                                      #Cell: 60 #Site: 100                                      \n",
            "                   CN_MIN: 1 CN_MAX: 5 ADO: 0.1 SEQ_ERR: 0.01 CN_NOISE: 0.05                    \n",
            "                           TREE_BATCH_SIZE: 128 NODE_BATCH_SIZE: 256                            \n",
            "───────────────────────────────────────── Local Search ─────────────────────────────────────────\n",
            "[04:50:31] [Iteration 0]   Likelihood: -12454.8754                                              \n",
            "[04:50:33] [Iteration 1]   Likelihood: -12448.8588                                              \n",
            "[04:50:35] [Iteration 2]   Likelihood: -12443.3693                                              \n",
            "[04:50:37] [Iteration 3]   Likelihood: -12440.0679                                              \n",
            "[04:50:40] [Iteration 4]   Likelihood: -12436.9292                                              \n",
            "[04:50:42] [Iteration 5]   Likelihood: -12431.9508                                              \n",
            "[04:50:44] [Iteration 6]   Likelihood: -12429.6368                                              \n",
            "[04:50:46] [Iteration 7]   Likelihood: -12427.5219                                              \n",
            "[04:50:48] [Iteration 8]   Likelihood: -12426.8218                                              \n",
            "[04:50:50] [Iteration 9]   Likelihood: -12423.0750                                              \n",
            "[04:50:53] [Iteration 10]  Likelihood: -12422.3967                                              \n",
            "[04:50:55] [Iteration 11]  Likelihood: -12421.7621                                              \n",
            "[04:50:57] [Iteration 12]  Likelihood: -12421.1561                                              \n",
            "[04:51:00] [Iteration 13]  Likelihood: -12420.5674                                              \n",
            "[04:51:02] [Iteration 14]  Likelihood: -12417.2646                                              \n",
            "[04:51:04] [Iteration 15]  Likelihood: -12414.3972                                              \n",
            "[04:51:06] [Iteration 16]  Likelihood: -12413.9839                                              \n",
            "[04:51:09] [Iteration 17]  Likelihood: -12413.6114                                              \n",
            "[04:51:11] [Iteration 18]  Likelihood: -12413.1327                                              \n",
            "[04:51:13] [Iteration 19]  Likelihood: -12412.8226                                              \n",
            "[04:51:15] [Iteration 20]  Likelihood: -12412.5275                                              \n",
            "[04:51:18] [Iteration 21]  Likelihood: -12412.2747                                              \n",
            "[04:51:20] [Iteration 22]  Likelihood: -12412.0630                                              \n",
            "[04:51:22] [Iteration 23]  Likelihood: -12411.6362                                              \n",
            "[04:51:24] [Iteration 24]  Likelihood: -12408.4896                                              \n",
            "[04:51:26] [Iteration 25]  Likelihood: -12408.1773                                              \n",
            "[04:51:29] [Iteration 26]  Likelihood: -12407.7867                                              \n",
            "[04:51:31] [Iteration 27]  Likelihood: -12407.6342                                              \n",
            "[04:51:33] [Iteration 28]  Likelihood: -12406.9226                                              \n",
            "[04:51:35] [Iteration 29]  Likelihood: -12406.0135                                              \n",
            "[04:51:37] [Iteration 30]  Likelihood: -12405.9710                                              \n",
            "[04:51:40] [Iteration 31]  Likelihood: -12405.9390                                              \n",
            "[04:51:42] [Iteration 32]  Likelihood: -12405.9105                                              \n",
            "[04:51:44] [Iteration 33]  Likelihood: -12405.8835                                              \n",
            "[04:51:47] Local search complete. Best Likelihood: -12405.883480072021                          \n",
            "────────────────────────────────────────────────────────────────────────────────────────────────\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Finally, we can print the inferred tree and genotypes."
      ],
      "metadata": {
        "id": "lxbEqdwyZ216"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print(scistreecna_tree)\n",
        "print(scistreecna_geno)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "h-Fiykz4aIBr",
        "outputId": "dabeb485-eaac-4d66-c6e3-0f6b38e17116"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(((((((((((((((c11,c16),c1),c38),c19),((c27,c49),c37)),(c0,c53)),((c23,c57),c41)),((((c15,c35),c29),(c24,c48)),((c17,c30),c59))),((c4,c45),c10)),c50),(((((c31,c7),c36),c25),(c52,c6)),((c14,c28),c58))),((((((((((c32,c56),c5),c22),(c20,c43)),(c55,c9)),c2),((c21,c51),c18)),(c33,c46)),c47),(c13,c54))),(c12,c34)),c42),((((c26,c44),c8),c3),(c39,c40)));\n",
            "[[0 0 0 ... 0 0 0]\n",
            " [0 0 0 ... 0 0 0]\n",
            " [0 0 0 ... 0 0 0]\n",
            " ...\n",
            " [0 1 0 ... 0 0 0]\n",
            " [0 0 0 ... 0 0 0]\n",
            " [0 0 0 ... 0 1 0]]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Since the ground truth for this dataset is available, we can load it and evaluate performance using two metrics:  \n",
        "- **Tree accuracy**, measured as $1 - Robinson\\_Foulds_{norm}$\n",
        "- **Genotype accuracy**\n"
      ],
      "metadata": {
        "id": "BI00pVtKaQQc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# load ground truth\n",
        "with open('./ScisTreeCNA/examples/test_data_tree.pkl', 'rb') as f:\n",
        "    true_tree = pickle.load(f)\n",
        "true_genotype = np.loadtxt('./ScisTreeCNA/examples/test_data_tg.txt', dtype=int)"
      ],
      "metadata": {
        "id": "VfKulWySaRW-"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "tree_accuracy = scna.util.tree_accuracy(true_tree, scistreecna_tree)\n",
        "geno_accuracy = scna.util.genotype_accuarcy(true_genotype, scistreecna_geno)\n",
        "print(f'Tree accuracy: {tree_accuracy}, Genotype accuracy: {geno_accuracy}')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rwJACX2SaXU9",
        "outputId": "452afeca-b595-4d9b-a98f-a84d29d520d3"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Tree accuracy: 0.4827586206896552, Genotype accuracy: 0.9911666666666666\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We can also run **scistree2** and compare the results with those from ScisTreeCNA.\n"
      ],
      "metadata": {
        "id": "tfziU1GRabqU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "scistree2tree, scistree2geno = scna.external.infer_scistree2_tree(reads, cell_names=cell_names)\n",
        "scistree2_tree_accuracy = scna.util.tree_accuracy(true_tree, scistree2tree)\n",
        "scistree2_geno_accuracy = scna.util.genotype_accuarcy(true_genotype, scistree2geno)\n",
        "print(f'Tree accuracy: {scistree2_tree_accuracy}, Genotype accuracy: {scistree2_geno_accuracy}')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6SaCUY_raczg",
        "outputId": "3cf55e8d-540e-4ca3-8943-d74660462b9e"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Tree accuracy: 0.3793103448275862, Genotype accuracy: 0.9893333333333333\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "(Optional) Actually, ScisTreeCNA integrates several external methods via `scn.external`, such as **CellPhy** and **DICE**.  \n",
        "These methods require users to install them separately and provide the corresponding executable files.  \n",
        "Here, we use **CellPhy** as an example.\n",
        "\n",
        ">Note: It is very slow to run CellPhy in colab as only a few CPU cores are available"
      ],
      "metadata": {
        "id": "shs3q5AUaf6y"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "We first install CellPhy, this may take a few minutes."
      ],
      "metadata": {
        "id": "hlTDUDJ9bbUd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%shell\n",
        "git clone https://github.com/amkozlov/cellphy\n",
        "cd cellphy\n",
        "sudo ./install.sh"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Slls13MFam29",
        "outputId": "b3aa00d2-b8f3-4817-afdc-78873c3721cf"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'cellphy'...\n",
            "remote: Enumerating objects: 196, done.\u001b[K\n",
            "remote: Counting objects: 100% (22/22), done.\u001b[K\n",
            "remote: Compressing objects: 100% (16/16), done.\u001b[K\n",
            "remote: Total 196 (delta 8), reused 19 (delta 6), pack-reused 174 (from 1)\u001b[K\n",
            "Receiving objects: 100% (196/196), 26.92 MiB | 13.29 MiB/s, done.\n",
            "Resolving deltas: 100% (95/95), done.\n",
            "Operating system detected: unknown\n",
            "Unknown OS: unknown\n",
            "Installing required R packages...\n",
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "trying URL 'https://cloud.r-project.org/src/contrib/BiocManager_1.30.27.tar.gz'\n",
            "Content type 'application/x-gzip' length 752490 bytes (734 KB)\n",
            "==================================================\n",
            "downloaded 734 KB\n",
            "\n",
            "* installing *source* package ‘BiocManager’ ...\n",
            "** this is package ‘BiocManager’ version ‘1.30.27’\n",
            "** package ‘BiocManager’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "*** copying figures\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (BiocManager)\n",
            "\n",
            "The downloaded source packages are in\n",
            "\t‘/tmp/RtmpBb5VOX/downloaded_packages’\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.2 (2025-10-31)\n",
            "Installing package(s) 'BiocVersion', 'ggtree'\n",
            "also installing the dependencies ‘fontBitstreamVera’, ‘fontLiberation’, ‘gridGraphics’, ‘fontquiver’, ‘ggplotify’, ‘patchwork’, ‘lazyeval’, ‘gdtools’, ‘ape’, ‘aplot’, ‘ggfun’, ‘yulab.utils’, ‘tidytree’, ‘treeio’, ‘ggiraph’\n",
            "\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/fontBitstreamVera_0.1.1.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/fontLiberation_0.1.0.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/gridGraphics_0.5-1.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/fontquiver_0.2.1.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/ggplotify_0.1.3.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/patchwork_1.3.2.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/lazyeval_0.2.2.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/gdtools_0.4.4.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/ape_5.8-1.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/aplot_0.2.9.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/ggfun_0.2.0.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/yulab.utils_0.2.1.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/tidytree_0.4.6.tar.gz'\n",
            "trying URL 'https://bioconductor.org/packages/3.22/bioc/src/contrib/treeio_1.34.0.tar.gz'\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/ggiraph_0.9.2.tar.gz'\n",
            "trying URL 'https://bioconductor.org/packages/3.22/bioc/src/contrib/BiocVersion_3.22.0.tar.gz'\n",
            "trying URL 'https://bioconductor.org/packages/3.22/bioc/src/contrib/ggtree_4.0.1.tar.gz'\n",
            "* installing *source* package ‘fontBitstreamVera’ ...\n",
            "** this is package ‘fontBitstreamVera’ version ‘0.1.1’\n",
            "** package ‘fontBitstreamVera’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** inst\n",
            "** help\n",
            "No man pages found in package  ‘fontBitstreamVera’ \n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (fontBitstreamVera)\n",
            "* installing *source* package ‘fontLiberation’ ...\n",
            "** this is package ‘fontLiberation’ version ‘0.1.0’\n",
            "** package ‘fontLiberation’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** inst\n",
            "** help\n",
            "No man pages found in package  ‘fontLiberation’ \n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (fontLiberation)\n",
            "* installing *source* package ‘gridGraphics’ ...\n",
            "** this is package ‘gridGraphics’ version ‘0.5-1’\n",
            "** package ‘gridGraphics’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (gridGraphics)\n",
            "* installing *source* package ‘patchwork’ ...\n",
            "** this is package ‘patchwork’ version ‘1.3.2’\n",
            "** package ‘patchwork’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "*** copying figures\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (patchwork)\n",
            "* installing *source* package ‘lazyeval’ ...\n",
            "** this is package ‘lazyeval’ version ‘0.2.2’\n",
            "** package ‘lazyeval’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** libs\n",
            "using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0’\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG       -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c expr.c -o expr.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG       -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c init.c -o init.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG       -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c interp.c -o interp.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG       -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c lazy.c -o lazy.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG       -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c name.c -o name.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG       -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c utils.c -o utils.o\n",
            "gcc -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o lazyeval.so expr.o init.o interp.o lazy.o name.o utils.o -L/usr/lib/R/lib -lR\n",
            "installing to /usr/local/lib/R/site-library/00LOCK-lazyeval/00new/lazyeval/libs\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** checking absolute paths in shared objects and dynamic libraries\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (lazyeval)\n",
            "* installing *source* package ‘ape’ ...\n",
            "** this is package ‘ape’ version ‘5.8-1’\n",
            "** package ‘ape’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** libs\n",
            "using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0’\n",
            "using C++ compiler: ‘g++ (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0’\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c BIONJ.c -o BIONJ.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c NNI.c -o NNI.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c RcppExports.cpp -o RcppExports.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c SPR.c -o SPR.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c additive.c -o additive.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c ape.c -o ape.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bNNI.c -o bNNI.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bionjs.c -o bionjs.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bipartition.c -o bipartition.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bitsplits.c -o bitsplits.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c delta_plot.c -o delta_plot.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c dist_dna.c -o dist_dna.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c dist_nodes.c -o dist_nodes.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c ewLasso.c -o ewLasso.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c heap.c -o heap.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c mat_expo.c -o mat_expo.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c me.c -o me.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c me_balanced.c -o me_balanced.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c me_ols.c -o me_ols.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c mvr.c -o mvr.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c mvrs.c -o mvrs.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c nj.c -o nj.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c njs.c -o njs.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c pic.c -o pic.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c plot_phylo.c -o plot_phylo.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c prop_part.cpp -o prop_part.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c rTrait.c -o rTrait.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c read_dna.c -o read_dna.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c reorder_Rcpp.cpp -o reorder_Rcpp.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c reorder_phylo.c -o reorder_phylo.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c treePop.c -o treePop.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c tree_build.c -o tree_build.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c tree_phylo.c -o tree_phylo.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c triangMtd.c -o triangMtd.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c triangMtds.c -o triangMtds.o\n",
            "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c ultrametric.c -o ultrametric.o\n",
            "g++ -std=gnu++17 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o ape.so BIONJ.o NNI.o RcppExports.o SPR.o additive.o ape.o bNNI.o bionjs.o bipartition.o bitsplits.o delta_plot.o dist_dna.o dist_nodes.o ewLasso.o heap.o mat_expo.o me.o me_balanced.o me_ols.o mvr.o mvrs.o nj.o njs.o pic.o plot_phylo.o prop_part.o rTrait.o read_dna.o reorder_Rcpp.o reorder_phylo.o treePop.o tree_build.o tree_phylo.o triangMtd.o triangMtds.o ultrametric.o -llapack -lblas -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n",
            "installing to /usr/local/lib/R/site-library/00LOCK-ape/00new/ape/libs\n",
            "** R\n",
            "** data\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** checking absolute paths in shared objects and dynamic libraries\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (ape)\n",
            "* installing *source* package ‘yulab.utils’ ...\n",
            "** this is package ‘yulab.utils’ version ‘0.2.1’\n",
            "** package ‘yulab.utils’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (yulab.utils)\n",
            "* installing *source* package ‘BiocVersion’ ...\n",
            "** this is package ‘BiocVersion’ version ‘3.22.0’\n",
            "** package ‘BiocVersion’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (BiocVersion)\n",
            "* installing *source* package ‘fontquiver’ ...\n",
            "** this is package ‘fontquiver’ version ‘0.2.1’\n",
            "** package ‘fontquiver’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (fontquiver)\n",
            "* installing *source* package ‘ggplotify’ ...\n",
            "** this is package ‘ggplotify’ version ‘0.1.3’\n",
            "** package ‘ggplotify’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (ggplotify)\n",
            "* installing *source* package ‘ggfun’ ...\n",
            "** this is package ‘ggfun’ version ‘0.2.0’\n",
            "** package ‘ggfun’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (ggfun)\n",
            "* installing *source* package ‘tidytree’ ...\n",
            "** this is package ‘tidytree’ version ‘0.4.6’\n",
            "** package ‘tidytree’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (tidytree)\n",
            "* installing *source* package ‘gdtools’ ...\n",
            "** this is package ‘gdtools’ version ‘0.4.4’\n",
            "** package ‘gdtools’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "Found pkg-config cflags and libs!\n",
            "Using PKG_CFLAGS=-I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid \n",
            "Using PKG_LIBS=-lcairo -lfreetype \n",
            "** libs\n",
            "using C++ compiler: ‘g++ (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0’\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid  -I../inst/include/ -I'/usr/lib/R/site-library/Rcpp/include'    -fvisibility=hidden -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c CairoContext.cpp -o CairoContext.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid  -I../inst/include/ -I'/usr/lib/R/site-library/Rcpp/include'    -fvisibility=hidden -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c RcppExports.cpp -o RcppExports.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid  -I../inst/include/ -I'/usr/lib/R/site-library/Rcpp/include'    -fvisibility=hidden -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c font_metrics.cpp -o font_metrics.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid  -I../inst/include/ -I'/usr/lib/R/site-library/Rcpp/include'    -fvisibility=hidden -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c gdtools.cpp -o gdtools.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid  -I../inst/include/ -I'/usr/lib/R/site-library/Rcpp/include'    -fvisibility=hidden -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c versions.cpp -o versions.o\n",
            "g++ -std=gnu++17 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o gdtools.so CairoContext.o RcppExports.o font_metrics.o gdtools.o versions.o -lcairo -lfreetype -L/usr/lib/R/lib -lR\n",
            "installing to /usr/local/lib/R/site-library/00LOCK-gdtools/00new/gdtools/libs\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** checking absolute paths in shared objects and dynamic libraries\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (gdtools)\n",
            "* installing *source* package ‘aplot’ ...\n",
            "** this is package ‘aplot’ version ‘0.2.9’\n",
            "** package ‘aplot’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "*** copying figures\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (aplot)\n",
            "* installing *source* package ‘treeio’ ...\n",
            "** this is package ‘treeio’ version ‘1.34.0’\n",
            "** package ‘treeio’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "*** copying figures\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (treeio)\n",
            "* installing *source* package ‘ggiraph’ ...\n",
            "** this is package ‘ggiraph’ version ‘0.9.2’\n",
            "** package ‘ggiraph’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** libs\n",
            "using C++ compiler: ‘g++ (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0’\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c RcppExports.cpp -o RcppExports.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c a_color.cpp -o a_color.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c clip.cpp -o clip.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c dsvg.cpp -o dsvg.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c dsvg_dev.cpp -o dsvg_dev.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c geom.cpp -o geom.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c indexed.cpp -o indexed.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c interactive.cpp -o interactive.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c mask.cpp -o mask.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c pattern.cpp -o pattern.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c raster.cpp -o raster.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c shapes.cpp -o shapes.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c svg.cpp -o svg.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c text.cpp -o text.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c tinyxml2.cpp -o tinyxml2.o\n",
            "g++ -std=gnu++17 -I\"/usr/share/R/include\" -DNDEBUG  -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/systemfonts/include'     -fpic  -g -O2 -ffile-prefix-map=/build/r-base-xupQTd/r-base-4.5.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -c utils.cpp -o utils.o\n",
            "g++ -std=gnu++17 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o ggiraph.so RcppExports.o a_color.o clip.o dsvg.o dsvg_dev.o geom.o indexed.o interactive.o mask.o pattern.o raster.o shapes.o svg.o text.o tinyxml2.o utils.o -lpng -lz -L/usr/lib/R/lib -lR\n",
            "installing to /usr/local/lib/R/site-library/00LOCK-ggiraph/00new/ggiraph/libs\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "*** copying figures\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** checking absolute paths in shared objects and dynamic libraries\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (ggiraph)\n",
            "* installing *source* package ‘ggtree’ ...\n",
            "** this is package ‘ggtree’ version ‘4.0.1’\n",
            "** package ‘ggtree’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** R\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** installing vignettes\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (ggtree)\n",
            "\n",
            "The downloaded source packages are in\n",
            "\t‘/tmp/RtmpBb5VOX/downloaded_packages’\n",
            "Done!\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": []
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Then we run CellPhy as follows."
      ],
      "metadata": {
        "id": "Cnlmko29cQV8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cellphytree, cellphygeno = scna.external.infer_cellphy_tree(reads, executable='./cellphy/cellphy.sh', num_threads=4)\n",
        "cellphy_tree_accuracy = scna.util.tree_accuracy(true_tree, cellphytree)\n",
        "cellphy_geno_accuracy = scna.util.genotype_accuarcy(true_genotype, cellphygeno)\n",
        "print(f'Tree accuracy: {cellphy_tree_accuracy}, Genotype accuracy: {cellphy_geno_accuracy}')"
      ],
      "metadata": {
        "id": "kYeMh7HUag7V"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}